Use of validation groups

It seems not possible to apply javax.validation annotation @ConvertGroup to REST service parameters when REST resource class implements an API Interface.

My REST resource class:
```
@RequestScoped
public class RestR implements IRestR {
    @Override
    @Path("path")
    @POST
    @GZIP
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public ResponseDTO postApi(
            @Valid @ConvertGroup(to = CreateCheck.class) @NotNull(groups = CreateCheck.class) InputDTO dto) {
            return null;
    }
}
```

It implements the following interface:
```
public interface IRestR {
    @Path("path")
    @POST
    @GZIP
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public ResponseDTO postApi(
            @Valid @ConvertGroup(to = CreateCheck.class) @NotNull(groups = CreateCheck.class) InputDTO dto
    );
}
```

The application cannot start since it generates this exception: `javax.validation.ConstraintDeclarationException: HV000124: Found multiple group conversions for source group javax.validation.groups.Default: com.myapp.api.dto.validation.groups.CreateCheck.`

Ommitting @ConvertGroup in API interface method would cause  `javax.validation.ConstraintDeclarationException: HV000151: A method overriding another method must not redefine the parameter constraint configuration, but method RestR#postApi(InputDTO) redefines the configuration of IRestR#postApi(InputDTO).`

Is there a way to exclude some annotations from being detected by redefinition check?

An alternative implementation of bean validation would require AspectJ that it is not supported by Quarkus.

Thanks @gsmet !

Unfortunately this error occurs when I try to build my project:
```
Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project myapp: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]     [error]: Build step io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor#build threw an exception: io.quarkus.deployment.configuration.ConfigurationError: Hibernate extension cannot guess the dialect as no database kind is specified by 'quarkus.datasource.db-kind'
[ERROR]     at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.guessDialect(HibernateOrmProcessor.java:857)
[ERROR]     at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.handleHibernateORMWithNoPersistenceXml(HibernateOrmProcessor.java:677)
[ERROR]     at io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor.build(HibernateOrmProcessor.java:237)
[ERROR]     at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]     at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR]     at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]     at java.lang.reflect.Method.invoke(Method.java:498)
[ERROR]     at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:931)
[ERROR]     at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR]     at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
[ERROR]     at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2027)
[ERROR]     at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1551)
[ERROR]     at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1442)
[ERROR]     at java.lang.Thread.run(Thread.java:748)
[ERROR]     at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

application.properties file includes the following lines:
```
quarkus.datasource.db-kind = postgresql
quarkus.hibernate-validator.method-validation.allow-overriding-parameter-constraints=true
```

--- a/docs/src/main/asciidoc/validation.adoc
+++ b/docs/src/main/asciidoc/validation.adoc
@@ -345,3 +345,7 @@ provided the supported locales have been properly specified in the `application.
 quarkus.locales=en-US,es-ES,fr-FR
 ----
 
+[[configuration-reference]]
+== Hibernate Validator Configuration Reference
+
+include::{generated-dir}/config/quarkus-hibernate-validator.adoc[leveloffset=+1, opts=optional]

--- a/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
+++ b/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
@@ -55,6 +55,7 @@ import io.quarkus.deployment.builditem.nativeimage.ReflectiveFieldBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveMethodBuildItem;
 import io.quarkus.deployment.logging.LogCleanupFilterBuildItem;
 import io.quarkus.deployment.recording.RecorderContext;
+import io.quarkus.hibernate.validator.runtime.HibernateValidatorBuildTimeConfig;
 import io.quarkus.hibernate.validator.runtime.HibernateValidatorRecorder;
 import io.quarkus.hibernate.validator.runtime.ValidatorProvider;
 import io.quarkus.hibernate.validator.runtime.interceptor.MethodValidationInterceptor;
@@ -89,6 +90,8 @@ class HibernateValidatorProcessor {
     private static final Pattern BUILT_IN_CONSTRAINT_REPEATABLE_CONTAINER_PATTERN = Pattern.compile("\\$List$");
 
     private LocalesBuildTimeConfig localesBuildTimeConfig;
+
+    private HibernateValidatorBuildTimeConfig hibernateValidatorBuildTimeConfig;
 
     @BuildStep
     HotDeploymentWatchedFileBuildItem configFile() {
@@ -245,7 +248,8 @@ class HibernateValidatorProcessor {
                                 hasXmlConfiguration(),
                                 capabilities.isCapabilityPresent(Capabilities.HIBERNATE_ORM),
                                 shutdownContext,
-                                localesBuildTimeConfig)));
+                                localesBuildTimeConfig,
+                                hibernateValidatorBuildTimeConfig)));
     }
 
     @BuildStep

--- a/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
+++ b/extensions/hibernate-validator/deployment/src/main/java/io/quarkus/hibernate/validator/deployment/HibernateValidatorProcessor.java
@@ -89,10 +89,6 @@ class HibernateValidatorProcessor {
 
     private static final Pattern BUILT_IN_CONSTRAINT_REPEATABLE_CONTAINER_PATTERN = Pattern.compile("\\$List$");
 
-    private LocalesBuildTimeConfig localesBuildTimeConfig;
-
-    private HibernateValidatorBuildTimeConfig hibernateValidatorBuildTimeConfig;
-
     @BuildStep
     HotDeploymentWatchedFileBuildItem configFile() {
         return new HotDeploymentWatchedFileBuildItem(META_INF_VALIDATION_XML);
@@ -146,7 +142,9 @@ class HibernateValidatorProcessor {
             BuildProducer<BeanContainerListenerBuildItem> beanContainerListener,
             ShutdownContextBuildItem shutdownContext,
             List<AdditionalJaxRsResourceMethodAnnotationsBuildItem> additionalJaxRsResourceMethodAnnotations,
-            Capabilities capabilities) throws Exception {
+            Capabilities capabilities,
+            LocalesBuildTimeConfig localesBuildTimeConfig,
+            HibernateValidatorBuildTimeConfig hibernateValidatorBuildTimeConfig) throws Exception {
 
         feature.produce(new FeatureBuildItem(FeatureBuildItem.HIBERNATE_VALIDATOR));
 

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/AllowMultipleCascadedValidationOnReturnValuesTest.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/AllowMultipleCascadedValidationOnReturnValuesTest.java
@@ -1,0 +1,66 @@
+package io.quarkus.hibernate.validator.test;
+
+import javax.inject.Inject;
+import javax.validation.Valid;
+import javax.validation.Validator;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class AllowMultipleCascadedValidationOnReturnValuesTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest().setArchiveProducer(() -> ShrinkWrap
+            .create(JavaArchive.class)
+            .addClasses(SubRealizationWithValidConstraintOnMethodParameter.class,
+                    RealizationWithValidConstraintOnMethodParameter.class,
+                    InterfaceWithNoConstraints.class)
+            .add(new StringAsset(
+                    "quarkus.hibernate-validator.method-validation.allow-multiple-cascaded-validation-on-return-values=true"),
+                    "application.properties"));
+
+    @Inject
+    Validator validator;
+
+    @Test
+    public void allowValidAddedInSubType() {
+        validator.forExecutables().validateParameters(
+                new SubRealizationWithValidConstraintOnMethodParameter(),
+                SubRealizationWithValidConstraintOnMethodParameter.class.getDeclaredMethods()[0],
+                new Object[] { "foo" });
+    }
+
+    private interface InterfaceWithNoConstraints {
+        String foo(String s);
+    }
+
+    private static class RealizationWithValidConstraintOnMethodParameter
+            implements InterfaceWithNoConstraints {
+        /**
+         * Adds @Valid to an un-constrained method from a super-type, which is not allowed.
+         */
+        @Override
+        @Valid
+        public String foo(String s) {
+            return "Hello Valid World";
+        }
+    }
+
+    private static class SubRealizationWithValidConstraintOnMethodParameter
+            extends RealizationWithValidConstraintOnMethodParameter {
+        /**
+         * Adds @Valid to an un-constrained method from a super-type, which is not allowed.
+         */
+        @Override
+        @Valid
+        public String foo(String s) {
+            return "Hello Valid World";
+        }
+    }
+
+}

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/AllowOverridingParameterConstraintsTest.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/AllowOverridingParameterConstraintsTest.java
@@ -1,0 +1,84 @@
+package io.quarkus.hibernate.validator.test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.Set;
+
+import javax.inject.Inject;
+import javax.validation.ConstraintViolation;
+import javax.validation.Validator;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class AllowOverridingParameterConstraintsTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest().setArchiveProducer(() -> ShrinkWrap
+            .create(JavaArchive.class)
+            .addClasses(RealizationWithMethodParameterConstraint.class, InterfaceWithNoConstraints.class,
+                    InterfaceWithNotNullMethodParameterConstraint.class,
+                    RealizationWithAdditionalMethodParameterConstraint.class)
+            .add(new StringAsset(
+                    "quarkus.hibernate-validator.method-validation.allow-overriding-parameter-constraints=true"),
+                    "application.properties"));
+
+    @Inject
+    Validator validator;
+
+    @Test
+    public void allowParameterConstraintsAddedInSubType() {
+        Set<? extends ConstraintViolation<?>> violations = validator.forExecutables().validateParameters(
+                new RealizationWithMethodParameterConstraint(),
+                RealizationWithMethodParameterConstraint.class.getDeclaredMethods()[0],
+                new Object[] { "foo" });
+
+        assertThat(violations).isEmpty();
+    }
+
+    @Test
+    public void allowStrengtheningInSubType() {
+        Set<ConstraintViolation<RealizationWithAdditionalMethodParameterConstraint>> violations = validator.forExecutables()
+                .validateParameters(
+                        new RealizationWithAdditionalMethodParameterConstraint(),
+                        RealizationWithAdditionalMethodParameterConstraint.class.getDeclaredMethods()[0],
+                        new Object[] { "foo" });
+
+        assertThat(violations).isEmpty();
+    }
+
+    private interface InterfaceWithNoConstraints {
+        String foo(String s);
+    }
+
+    private static class RealizationWithMethodParameterConstraint implements InterfaceWithNoConstraints {
+        /**
+         * Adds constraints to an un-constrained method from a super-type, which is not allowed.
+         */
+        @Override
+        public String foo(@NotNull String s) {
+            return "Hello World";
+        }
+    }
+
+    private interface InterfaceWithNotNullMethodParameterConstraint {
+        void bar(@NotNull String s);
+    }
+
+    private static class RealizationWithAdditionalMethodParameterConstraint
+            implements InterfaceWithNotNullMethodParameterConstraint {
+        /**
+         * Adds constraints to a constrained method from a super-type, which is not allowed.
+         */
+        @Override
+        public void bar(@Size(min = 3) String s) {
+        }
+    }
+}

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/AllowParameterConstraintsOnParallelMethodsTest.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/AllowParameterConstraintsOnParallelMethodsTest.java
@@ -1,0 +1,55 @@
+package io.quarkus.hibernate.validator.test;
+
+import javax.inject.Inject;
+import javax.validation.Validator;
+import javax.validation.constraints.NotNull;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class AllowParameterConstraintsOnParallelMethodsTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest().setArchiveProducer(() -> ShrinkWrap
+            .create(JavaArchive.class)
+            .addClasses(RealizationOfTwoInterface.class, InterfaceWithNoConstraints.class,
+                    AnotherInterfaceWithMethodParameterConstraint.class)
+            .add(new StringAsset(
+                    "quarkus.hibernate-validator.method-validation.allow-parameter-constraints-on-parallel-methods=true"),
+                    "application.properties"));
+
+    @Inject
+    Validator validator;
+
+    @Test
+    public void allowParameterConstraintsInHierarchyWithMultipleRootMethods() {
+        validator.forExecutables().validateParameters(
+                new RealizationOfTwoInterface(),
+                RealizationOfTwoInterface.class.getDeclaredMethods()[0],
+                new Object[] { "foo" });
+    }
+
+    private interface InterfaceWithNoConstraints {
+        String foo(String s);
+    }
+
+    private interface AnotherInterfaceWithMethodParameterConstraint {
+        String foo(@NotNull String s);
+    }
+
+    private static class RealizationOfTwoInterface
+            implements InterfaceWithNoConstraints, AnotherInterfaceWithMethodParameterConstraint {
+        /**
+         * Implement a method that is declared by two interfaces, one of which has a constraint
+         */
+        @Override
+        public String foo(String s) {
+            return "Hello World";
+        }
+    }
+}

--- a/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/FailFastTest.java
+++ b/extensions/hibernate-validator/deployment/src/test/java/io/quarkus/hibernate/validator/test/FailFastTest.java
@@ -1,0 +1,66 @@
+package io.quarkus.hibernate.validator.test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.inject.Inject;
+import javax.validation.ConstraintViolation;
+import javax.validation.Valid;
+import javax.validation.Validator;
+import javax.validation.constraints.Email;
+import javax.validation.constraints.Max;
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Pattern;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class FailFastTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest().setArchiveProducer(() -> ShrinkWrap
+            .create(JavaArchive.class)
+            .addClasses(A.class, B.class)
+            .add(new StringAsset("quarkus.hibernate-validator.fail-fast=true"),
+                    "application.properties"));
+
+    @Inject
+    Validator validator;
+
+    @Test
+    public void testFailFastSetOnConfiguration() {
+        Set<ConstraintViolation<A>> constraintViolations = validator.validate(new A());
+
+        assertThat(constraintViolations).hasSize(1);
+    }
+
+    class A {
+        @NotNull
+        String b;
+
+        @NotNull
+        @Email
+        String c;
+
+        @Pattern(regexp = ".*\\.txt$")
+        String file;
+
+        @Valid
+        Set<B> bs = new HashSet<B>();
+    }
+
+    class B {
+        @Min(value = 10)
+        @Max(value = 30)
+        @NotNull
+        Integer size;
+    }
+}

--- a/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorBuildTimeConfig.java
+++ b/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorBuildTimeConfig.java
@@ -1,0 +1,75 @@
+package io.quarkus.hibernate.validator.runtime;
+
+import io.quarkus.runtime.annotations.ConfigDocSection;
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
+public class HibernateValidatorBuildTimeConfig {
+
+    /**
+     * Enable the fail fast mode. When fail fast is enabled the validation
+     * will stop on the first constraint violation detected.
+     */
+    @ConfigItem(defaultValue = "false")
+    public boolean failFast;
+
+    /**
+     * Method validation.
+     */
+    @ConfigDocSection
+    public HibernateValidatorMethodBuildTimeConfig methodValidation;
+
+    @ConfigGroup
+    public static class HibernateValidatorMethodBuildTimeConfig {
+
+        /**
+         * Define whether overriding methods that override constraints should throw a {@code ConstraintDefinitionException}.
+         * The default value is {@code false}, i.e. do not allow.
+         * <p>
+         * See Section 4.5.5 of the JSR 380 specification, specifically
+         *
+         * <pre>
+         * "In sub types (be it sub classes/interfaces or interface implementations), no parameter constraints may
+         * be declared on overridden or implemented methods, nor may parameters be marked for cascaded validation.
+         * This would pose a strengthening of preconditions to be fulfilled by the caller."
+         * </pre>
+         */
+        @ConfigItem(defaultValue = "false")
+        public boolean allowOverridingParameterConstraints;
+
+        /**
+         * Define whether parallel methods that define constraints should throw a {@code ConstraintDefinitionException}. The
+         * default value is {@code false}, i.e. do not allow.
+         * <p>
+         * See Section 4.5.5 of the JSR 380 specification, specifically
+         *
+         * <pre>
+         * "If a sub type overrides/implements a method originally defined in several parallel types of the hierarchy
+         * (e.g. two interfaces not extending each other, or a class and an interface not implemented by said class),
+         * no parameter constraints may be declared for that method at all nor parameters be marked for cascaded validation.
+         * This again is to avoid an unexpected strengthening of preconditions to be fulfilled by the caller."
+         * </pre>
+         */
+        @ConfigItem(defaultValue = "false")
+        public boolean allowParameterConstraintsOnParallelMethods;
+
+        /**
+         * Define whether more than one constraint on a return value may be marked for cascading validation are allowed.
+         * The default value is {@code false}, i.e. do not allow.
+         * <p>
+         * See Section 4.5.5 of the JSR 380 specification, specifically
+         *
+         * <pre>
+         * "One must not mark a method return value for cascaded validation more than once in a line of a class hierarchy.
+         * In other words, overriding methods on sub types (be it sub classes/interfaces or interface implementations)
+         * cannot mark the return value for cascaded validation if the return value has already been marked on the
+         * overridden method of the super type or interface."
+         * </pre>
+         */
+        @ConfigItem(defaultValue = "false")
+        public boolean allowMultipleCascadedValidationOnReturnValues;
+    }
+}

--- a/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorRecorder.java
+++ b/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateValidatorRecorder.java
@@ -32,7 +32,8 @@ public class HibernateValidatorRecorder {
     public BeanContainerListener initializeValidatorFactory(Set<Class<?>> classesToBeValidated,
             Set<String> detectedBuiltinConstraints,
             boolean hasXmlConfiguration, boolean jpaInClasspath,
-            ShutdownContext shutdownContext, LocalesBuildTimeConfig localesBuildTimeConfig) {
+            ShutdownContext shutdownContext, LocalesBuildTimeConfig localesBuildTimeConfig,
+            HibernateValidatorBuildTimeConfig hibernateValidatorBuildTimeConfig) {
         BeanContainerListener beanContainerListener = new BeanContainerListener() {
 
             @Override
@@ -99,6 +100,14 @@ public class HibernateValidatorRecorder {
                 }
 
                 // Hibernate Validator-specific configuration
+
+                configuration.failFast(hibernateValidatorBuildTimeConfig.failFast);
+                configuration.allowOverridingMethodAlterParameterConstraint(
+                        hibernateValidatorBuildTimeConfig.methodValidation.allowOverridingParameterConstraints);
+                configuration.allowParallelMethodsDefineParameterConstraints(
+                        hibernateValidatorBuildTimeConfig.methodValidation.allowParameterConstraintsOnParallelMethods);
+                configuration.allowMultipleCascadedValidationOnReturnValues(
+                        hibernateValidatorBuildTimeConfig.methodValidation.allowMultipleCascadedValidationOnReturnValues);
 
                 InstanceHandle<ScriptEvaluatorFactory> configuredScriptEvaluatorFactory = Arc.container()
                         .instance(ScriptEvaluatorFactory.class);

